Entropy: nRF: Clear NVIC flags before CPU idling

ioannisg opened the pull request:
Fixes #26105 

Aims at fixing an (ugly) race condition in the entropy driver.

Tested on
- nrf51dk_nrf51422
- nrf52dk_nrf52832
- nrf52833dk_nrf52833
- nrf52840dk_nrf52840

ioannisg committed:
drivers: entropy: nRF: clear NVIC pending before doing WFE on RNG IRQ

We need to clear the NVIC Pending bit for the RNG IRQ before
doing any WFEs and expect to wake up by RNG events. This is
because the event register will be set only if NVIC status
is changed from 0 to 1.

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

ioannisg committed:
drivers: entropy: nrf5: add docs and a DSB in get_entropy_isr

Add a DSB before doing WFE in get_entropy_isr, to ensure
the memory transactions are complete.

Add a note clarifying a dependency for the existing solution
(dependency is satisfied by ARCH code but is good to state
clearly).

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

pabigot commented on drivers/entropy/entropy_nrf5.c:
```diff
@@ -278,6 +278,12 @@ static int entropy_nrf5_get_entropy_isr(struct device *dev, uint8_t *buf, uint16
+		 * will set the Cortex-M single-bit event register to 1 (the bit is
+		 * set when NVIC pending IRQ status is changed from 0 to 1)
+		 */
+		NVIC_ClearPendingIRQ(IRQN);
```
Should this be before the trigger?

ioannisg replied:
Well, that's probably the "intuitive thing to do", although it, clearly, has no impact. That's because: 
- the task could be already triggered but a previous invocation - we make no assumptions the task is triggered here for the first time
- as long as it _becomes_ clear _after_ we have triggered the task and _before_ we enter the while loop containing WFE, we are certain that the event will occur

No strong opinion though

pabigot replied:
OK

ioannisg commented:
@aescolar @wopu-ot could you assist here? I provided a very simple model for ARM's DSB intrinsic as "no-op" for nrf52_bsim, I wonder if this should be sufficient. Effectively a DSB makes sure memory writes are complete before executing next instruction, so it should not matter for the simulated platform (but would like you to take a look)

cvinayak reviewed (approved):
LGTM

wopu-ot commented on boards/posix/nrf52_bsim/irq_handler.c:
```diff
@@ -390,3 +390,8 @@ void __SEV(void)
 }
+
+/* Model ARM Data Synchronization Barrier instruction as no-op */
+void __DSB(void)
```
The no-op should be sufficient for nrf52_bsim. __DMB is defined in the HW model (see https://github.com/BabbleSim/ext_NRF52_hw_models/blob/master/src/nrfx/hal/nrf_soc_if.h#L98) . For consistency, I would prefer __DSB to come from there too.

aescolar replied:
I agree with @wopu-ot that this one does not belong here (the `__WFE` `__WFI` are here because of the model of the event behaviour)
But I'm thinking that maybe the `__DMB` and `__NOP` should not have been placed in the HW models, but in a new `boards/posix/nrf52_bsim/cmsis.h` as macros, and have this `cmsis.h` be included from `board_soc.h`. (This __<asm_inst> macros/functions are defined by CMSIS)

@wopu-ot what do you think about also moving the DMB and NOP to a new cmsis.h?

ioannisg replied:
@aescolar @wopu-ot thanks for your input on this. I would probably prefer to see the functions in some cmsis.h as macros or static inline empty functions, what do you think?

wopu-ot replied:
@aescolar @ioannisg I would be fine with moving DMB and NOP, my point was more that DSB should be next to DMB.

aescolar replied:
@ioannisg Would you then create that new file with these definitions, and send a PR to the HW models removing those 2 static inlines from there?
@rugeGerritsen I guess you are ok with __DMB and __NOP being in that new cmsis.h?

rugeGerritsen replied:
@aescolar . I'm fine with moving those to cmsis.h

ioannisg replied:
@aescolar do get this PR pass CI, I suggest I do this for DSB and we can move the DMP, NOP, in the cmisis (and remove from BabbleSIm) in a subsequent PR, what do you think?

aescolar replied:
@ioannisg it can be. Would it be you the kind person preparing that 2nd PR? :)

ioannisg replied:
sure, can do that

ioannisg replied:
@aescolar @wopu-ot https://github.com/BabbleSim/ext_NRF52_hw_models/pull/6
I guess, if you merge that one, then we can send a PR in Zephyr to add these functions in cmsis.h

ioannisg committed:
boards: posix: nrf52_sim: model DSB as no-op in cmsis.h

Provide a very simple model of the DSB ARM
instruction as no-operation. The implementation
is provided in a cmsis.h header file added
in the nrf52_bsim board directory.

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

pabigot reviewed (commented):
No objections, but it's far enough out of of my comfort zone that I don't feel qualified to give an explicit approval.

carlescufi commented on drivers/entropy/entropy_nrf5.c:
```diff
@@ -278,11 +278,27 @@ static int entropy_nrf5_get_entropy_isr(struct device *dev, uint8_t *buf, uint16
 		nrf_rng_event_clear(NRF_RNG, NRF_RNG_EVENT_VALRDY);
 		nrf_rng_task_trigger(NRF_RNG, NRF_RNG_TASK_START);
 
+		/* Clear NVIC pending bit. This ensures that a subsequent
+		 * RNG event will set the Cortex-M single-bit event register
+		 * to 1 (the bit is set when NVIC pending IRQ status is
+		 * changed from 0 to 1)
+		 */
+		NVIC_ClearPendingIRQ(IRQN);
```
So what happens if the ISR executes between these 2? It doesn't matter because we have the `nrf_rng_event_check()` later?

ioannisg replied:
@carlescufi the problem we try to solve here is the following: 
When the `get_entropy_isr()` function is called with the RNG interrupt masked, that is
- either IRQs are (globally)  locked, or
- the RNG IRQ is not enabled, or
- the function is called from a higher handler priority

then it is possible that the NVIC pending flag of the RNG interrupt is already set (to 1) before we do the first WFE. This means that the ARM core won't set the single-bit event register when the next RNG number is ready, since this is done when the NVIC pending IRQ flag switches from 0 to 1. This event register might be set due to other reasons, of course, but **if it is not, then the MCU cannot exit WFE when the next RNG number is ready**

> So what happens if the ISR executes between these 2?

Actually, the IRQ is disabled on line 274, so it cannot execute between the clear and the trigger. But it can be set to pending, which will set the event register and allow us to exit WFE. 

To summarize:
By adding an NVIC_ClearPendingIRQ(IRQN); statement before the first WFE, I guarantee that we will always witness a pending flag switch from 0 to 1, and this event will set the event-register bit to 1, allowing to exit WFE.

ioannisg replied:
@stephanosio maybe you could also take a quick look and review this reasoning?
